imap/index.c

```diff
@@ -3813,6 +3813,124 @@ static int index_fetchannotations(struct index_state *state,
     return r;
 }
 
+struct fetch_mailbox_rock {
+    struct index_state *state;
+    int sep;
+    int wantname;
+    const struct fetchargs *fetchargs;
+};
+
+static int fetch_mailbox_cb(const conv_guidrec_t *rec, void *rock)
+{
+    static const int needrights = ACL_READ|ACL_LOOKUP;
+    struct fetch_mailbox_rock *fmb_rock = (struct fetch_mailbox_rock *) rock;
+    mbentry_t *mbentry = NULL;
+    int myrights = 0;
+    struct mailbox *mailbox = NULL;
+    msgrecord_t *msgrecord = NULL;
+    char *extname = NULL;
+    uint32_t system_flags, internal_flags;
+    int r = 0;
+
+    assert(fmb_rock->state != NULL);
+    assert(fmb_rock->fetchargs != NULL);
+
+    /* make sure we have appropriate rights */
+    r = mboxlist_lookup(rec->mboxname, &mbentry, NULL);
+    if (r) goto done;
+    myrights = cyrus_acl_myrights(fmb_rock->state->authstate, mbentry->acl);
+    if ((myrights & needrights) != needrights)
+        goto done;
+
+    if (rec->version >= 1) {
```

Comment:
Nit: Checking version >= 1 and expunged/deleted flags at the start of this function would allow to skip a mboxlist and ACL lookup call for expunged messages. At least for FastMail or other recent installations that upgrade their conversations.db, the version will always be >= 1 soon.